packages/vite/src/node/plugins/clientInjections.ts

```diff
@@ -47,7 +50,7 @@ export function clientInjectionsPlugin(config: ResolvedConfig): Plugin {
           .replace(`__DEFINES__`, serializeDefine(config.define || {}))
           .replace(`__HMR_PROTOCOL__`, JSON.stringify(protocol))
           .replace(`__HMR_HOSTNAME__`, JSON.stringify(host))
-          .replace(`__HMR_PORT__`, JSON.stringify(port))
+          .replace(`__HMR_PORT__`, options.runtimePort ?? JSON.stringify(port))
```

Comment:
With the changes above, this no longer needs to be changed:

```suggestion
          .replace(`__HMR_PORT__`, JSON.stringify(port))
```